### PR TITLE
reversible dynamics test

### PR DESCRIPTION
**Pull request by davidBelanger**
This implements a simple example of the work in  Maclaurin, Duvenaud, and Adams. "Gradient-based Hyperparameter Optimization through Reversible Learning." We learn a linear regression model with SGD, and we differentiate the test set performance of the model with respect to the learning rate used at train time. 
Right now, the code crashes with the error: 

Optimized_ReversibleDynamics
 Function call failed 
...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:375: ...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:56: attempt to index global 'output' (a nil value)
stack traceback:
	[C]: in function 'error'
	...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:375: in function 'record'
	.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:44: in function 'generateFn'
	.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:93: in function 'dLearn'
	test/test.lua:1639: in function <test/test.lua:1547>
	[C]: in function 'xpcall'
	...rs/belanger/torch/install/share/lua/5.1/totem/Tester.lua:556: in function '_pcall'
	...rs/belanger/torch/install/share/lua/5.1/totem/Tester.lua:853: in function '_run'
	...rs/belanger/torch/install/share/lua/5.1/totem/Tester.lua:727: in function '_runCL'
	...rs/belanger/torch/install/share/lua/5.1/totem/Tester.lua:747: in function 'run'
	test/test.lua:1658: in main chunk
	[C]: in function 'dofile'
	...nger/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
	[C]: at 0x010d67cc40

### CODE REVIEW

**Comment by luketwitter** [comment]
There were a few more system bugs with grads of grads that I've fixed.

Also, the test has a few errors, specifically:

- torch.Tensor() syntax doesn't work yet. Use torch.Tensor.new().
- The training and data sets must be passed as (non-differentiable) params.
- We don't yet support the 4 argument torch.add, or any of the torch.XXX functions that operate in place. I've written it as: params.W = params.W + grads.W * -metaParams.learningRate

With all that fixed, it gets a lot further. Let me know if the values being produced are nonsense or not.

```

   ReversibleDynamics = function()
      --This implements a very simple example of the concepts in:
      --Maclaurin, Duvenaud, and Adams. "Gradient-based Hyperparameter Optimization through Reversible Learning."
      --We learn a linear regression model with SGD, and differentiate the test set performance of the model with respect to the learning rate used at train time
      --Ideally this would be used to learn the learning rate using gradient descent. This test is more simple.
      --It just evaluates the gradient of the loss with respect to the learningRate for a few different learning rate values.
      --Also, Maclaurin et al. use SGD with momentum, but this test uses simple SGD.

      -- define trainable parameters:
      local numFeatures = 10
      local numOutputs = 3
      local numSamples = 250
      local numEpochs = 1

      --this sets up the true model from which the data is drawn
      local trueParams = {
         W = torch.randn(numFeatures,numOutputs),
         b = torch.randn(numOutputs)
      }

      --this is used to generate train and test datasets
      local function genDataset()
         local dataset = {}
         for tt = 1,numSamples do
            local x = torch.randn(1,numFeatures)
            local y = x*trueParams.W + trueParams.b
            table.insert(dataset,{x = x,y = y})
         end
         return dataset
      end
      local trainingSet = genDataset()
      local testSet = genDataset()


      -- define loss for linear regression model
      local linearRegressionLoss = function(params, x, y)
         local yhat = x*params.W + params.b
         local residual = y - yhat
         local loss = 0.5*torch.sum(torch.cmul(residual,residual))
         return loss
      end

      -- gradient of loss
      local dLinearRegressionLoss = autograd(linearRegressionLoss)


      --we learn the regression parameters using SGD with a fixed step size
      function gradientUpdate(metaParams,params,grads)
         params.W = params.W + grads.W * -metaParams.learningRate
         params.b = params.b + grads.b * -metaParams.learningRate
      end

      --this initializes the params that are optimized by learning.
      function getInitParams()
         local p =  {
            W = torch.zero(torch.Tensor.new(numFeatures,numOutputs)),
            b = torch.zero(torch.Tensor.new(numOutputs))
         }
         return p
      end

      --this learns a regression model on the train set and evaluates its squared loss on the test set
      --its behavior depends on the value of learningRate in metaParams
      function learn(metaParams, trainingSet, testSet)
         local params = getInitParams()
         for epoch = 1,numEpochs do
            for _,sample in ipairs(trainingSet) do
               local grads, loss = dLinearRegressionLoss(params, sample.x, sample.y)
               gradientUpdate(metaParams,params,grads)
            end
         end

         local testLoss = torch.zero(torch.Tensor(1))
         local numSamples = #testSet
         for _,sample in ipairs(testSet) do
            local g, tL = dLinearRegressionLoss(params,sample.x,sample.y)
            testLoss = testLoss + tL
         end
         return testLoss[1]
      end

      --this is the gradient of the test loss with respect to the learning rate used on the train data
      local dLearn = autograd(learn, { inline = false })

      --this just evaluates dLearn for a few choices of the learning rate. Eventually, we would like to learn the learningRate with gradient descent.
      for _,lr in ipairs({0.001,0.01,0.1}) do

         --this just evaluates the 'learn' function directly, in order to demonstrate that this code path works
         local testLoss1 = learn({learningRate = lr}, trainingSet, testSet)

         --this evaluates the derivative of the 'learn function.' Currently it is failing.
         local testLoss2, dMetaParams = dLearn({learningRate = lr}, trainingSet, testSet)
         print(testLoss2)
         tester:assert(testLoss1 == testLoss2,"value computed using 'learn' function does not match value computed by 'dLearn'")
      end
   end
```

**Comment by davidBelanger** [comment]
Thanks for the help. I currently have it set up to closely mirror the paper
<http://arxiv.org/pdf/1502.03492.pdf>. It is working. I can tune my
hyperparameters by directly optimizing for dev set performance.
Unfortunately, if I scale things up I hit numerical overflow/underflow
issues, as discussed in sections 2.2 and 2.3 of the paper.

If you'd like to use the code as an example, I'd be happy to make a PR for
it in an examples directory or something. Unfortunately, it only works for
small problems, though. I'm closing the PR for now.

**Comment by davidBelanger** [comment]
To clarify, my fork has an updated version of the test:
https://github.com/twitter/torch-autograd/blob/master/test/test.lua
If you'd like me to open an new PR for  that, I'd be happy to. My concern is that it's hard to make a well-defined test for it since it's hard to check that values have overflowed (they aren't nan or inf; they're just very big values).
